File: 417-pacific-atlantic-water-flow/pacific-atlantic-water-flow.cpp
class Solution {
    vector<pair<int, int>> directions = {{1, 0}, {-1, 0},
                                         {0, 1}, {0, -1}};
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {

        int rows=heights.size();
        int cols=heights[0].size();




        vector<vector<bool>>pac(rows,vector<bool>(cols, false));
        vector<vector<bool>>alt(rows,vector<bool>(cols, false));

        for(int c=0;c<cols;c++)
        {
            dfs(0,c,pac,heights);
            dfs(rows-1,c,alt,heights);
        }

        for(int r=0;r<rows;r++)
        {
            dfs(r,0,pac, heights);
            dfs(r,cols-1,alt, heights);



        }
        vector<vector<int>>res;

        for(int r=0;r<rows;r++)
        {
            for(int c=0;c<cols;c++)
            {
                if(pac[r][c] && alt[r][c]) 
                {
                    res.push_back({r,c});
                }
            }
        }

        return res;





        
    }

    void dfs(int r, int c,vector<vector<bool>>&ocean,vector<vector<int>>& heights)
    {

        ocean[r][c]=true;

        for(auto [dr,dc]: directions)
        {
            int nr=r+dr;
            int nc=c+dc;

            if(nr>=0 && nr<heights.size() && nc>=0 && nc<heights[0].size() && heights[nr][nc]>=heights[r][c] && !ocean[nr][nc])
            {
                dfs(nr,nc,ocean,heights);
            }

        }

       
    }
};

File: 417-pacific-atlantic-water-flow/pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        rows=len(heights)
        cols=len(heights[0])


        atl=[[False]*cols for _ in range(rows)]
        pac= [[False]*cols for _ in range(rows)]


        # now start from any cell and traverse and mark it true only if the neihboring
        # cell has height more than current one
        def dfs(row,col,ocean):
            ocean[row][col]=True

            for dr,dc in [(1,0),(-1,0),(0,1),(0,-1)]:
                nr=row+dr
                nc=col+dc

                if 0<=nr<rows and 0<=nc<cols and not ocean[nr][nc] and heights[nr][nc]>=heights[row][col]:
                    dfs(nr,nc,ocean)

        for c in range(cols):

            dfs(0,c,pac)
            dfs(rows-1,c,atl)
        for r in range(rows):

            dfs(r,0,pac)
            dfs(r,cols-1,atl)
        res=[]
        for r in range(rows):
            for c in range(cols):
                if atl[r][c] and pac[r][c]:
                    res.append([r,c])

        return res

        